server/pulp/server/async/tasks.py

```diff
@@ -373,20 +452,110 @@ def serialize(self):
+
+         See the docblock on _queue_reserved_task and _queue_reserved_task_list for more
+         information.
+
+         This method creates a TaskStatus as a placeholder for later updates. Pulp expects to poll
+         on a task just after calling this method, so a TaskStatus entry needs to exist for it
+         before it returns.
+
+         For a list of parameters accepted by the *args and **kwargs parameters, please see the
+         docblock for the apply_async() method.
+
+         :param reservation:    A string-or-list-of-strings that identify a set of named resources,
+                                guaranteeing that only one task reserving any resource-ids in this
+                                list can happen at a time.
+         :type  reservation:    list or basestr
+         :param tags:           A list of tags (strings) to place onto the task, used for searching
+                                for tasks by tag
+         :type  tags:           list
+         :param group_id:       The id to identify which group of tasks a task belongs to
+         :type  group_id:       uuid.UUID
+         :return:               An AsyncResult instance as returned by Celery's apply_async
+         :rtype:                celery.result.AsyncResult
+         """
+        inner_task_id = str(uuid.uuid4())
+        task_name = self.name
+        tag_list = kwargs.get('tags', [])
+        group_id = kwargs.get('group_id', None)
+
+        # Create a new task status with the task id and tags.
+        task_status = TaskStatus(task_id=inner_task_id, task_type=task_name,
+                                 state=constants.CALL_WAITING_STATE, tags=tag_list,
+                                 group_id=group_id)
+        # To avoid the race condition where __call__ method below is called before
+        # this change is propagated to all db nodes, using an 'upsert' here and setting
+        # the task state to 'waiting' only on an insert.
+        task_status.save_with_set_on_insert(fields_to_set_on_insert=['state', 'start_time'])
+        try:
+            # Decide what to call based on the kind-of reservation(s) we are being asked to make
+            if isinstance(reservation, list):
+                _queue_reserved_task_list.apply_async(
+                    args=[task_name, inner_task_id, reservation, args, kwargs],
+                    queue=RESOURCE_MANAGER_QUEUE
+                )
+            else:
+                _queue_reserved_task.apply_async(
+                    args=[task_name, inner_task_id, reservation, args, kwargs],
+                    queue=RESOURCE_MANAGER_QUEUE
+                )
+        except Exception:
+            TaskStatus.objects(task_id=task_status.task_id).update(state=constants.CALL_ERROR_STATE)
+            raise
+
+        return AsyncResult(inner_task_id)
+
+    def apply_async_with_reservation_list(self, resource_type, resource_id_list, *args, **kwargs):
```

Comment:
I'm not convinced we should make another dispatch interface with the `_list` name. What if Can't we generalize the existing one to coerce the single type case to a list with something like:

```
if type(resource_id) != 'list':
    list(resource_id)
```

What do you think about something like that?

Reply:
if we do that, then will we assume that all the resources are of the same type? Or will we assume that the resource type is a list also and each item corresponds to an item in the list of resource ids?

Reply:
I went back and forth about that with myself and w/@dkliban. Making a new API makes it clearer where we're grabbing multiple resource locks. Also makes the proposal above (list-of-tuples) possible, without the confusion of "why do I have to provide a resource-type arg when I built all of them into the list I'm passing in?"
I mean, the *cleanest* thing to do is have one API that takes a list-of-tuples - but my first concern w/Pulp2 is "do no harm", and I didn't think going back and finding/changing all the existing code (which works) to use a new API just to be marginally cleaner, was a good choice - more risk, low return. So instead opted to be explicit about what's going on, when we call this API, and leave the existing code as unchanged as I could. Make any sense?

Reply:
@dkliban I had assumed the latter. That is how Pulp3 works at least. Resource is a list of strings:

resource_id_list = ['lock1', 'lock2', 'lock3']

What is written here is ok too. I think I just imagined it simpler with one interface.

Reply:
async_with_reservation took a type and an id and made a combined-key out of them, in order to essentially namespace the reservations, so _list needs to take a list of (type,id) tuples to accomplish the same goal - since we don't want to push the combined-key-creation out onto the caller.

Reply:
@ggainey I see. Thank you. +1 to the list of tuples described here https://github.com/pulp/pulp/pull/3945/files#r307866063